File: kora-java-kafka/src/main/java/ru/tinkoff/kora/example/kafka/listener/AbstractListener.java
package ru.tinkoff.kora.example.kafka.listener;

import java.util.ArrayList;
import java.util.List;

public abstract class AbstractListener<T> {

    private final List<T> events = new ArrayList<>();
    private final List<Exception> exceptions = new ArrayList<>();

    void success(T event) {
        events.add(event);
    }

    void fail(Exception exception) {
        exceptions.add(exception);
    }

    public List<T> received() {
        return events;
    }

    public List<Exception> failed() {
        return exceptions;
    }

    public void reset() {
        events.clear();
        exceptions.clear();
    }
}

File: examples/kotlin/kora-kotlin-kafka/src/main/kotlin/ru/tinkoff/kora/kotlin/example/kafka/listener/AbstractListener.kt
package ru.tinkoff.kora.kotlin.example.kafka.listener

abstract class AbstractListener<T> {
    private val events = ArrayList<T>()
    private val exceptions = ArrayList<Exception>()

    fun success(event: T) {
        events.add(event)
    }

    fun fail(exception: Exception) {
        exceptions.add(exception)
    }

    fun received(): List<T> = events

    fun failed(): List<Exception> = exceptions

    fun reset() {
        events.clear()
        exceptions.clear()
    }
}
